drivers/dac/ad5754r/ad5754r.c

```diff
@@ -0,0 +1,684 @@
+				      init_param->gpio_bin_2scomp_init);
+	if (ret)
+		return ret;
+
+	if (dev->gpio_bin_2scomp) {
+		ret = no_os_gpio_direction_output(dev->gpio_bin_2scomp,
+						  NO_OS_GPIO_HIGH);
+		if (ret)
+			return ret;
+	}
+
+	/* GPIO Clear Initialization */
+	ret = no_os_gpio_get_optional(&dev->gpio_clear,
+				      init_param->gpio_clear_init);
+	if (ret)
+		return ret;
+
+	if (dev->gpio_clear) {
+		ret = no_os_gpio_direction_output(dev->gpio_clear,
+						  NO_OS_GPIO_HIGH);
+		if (ret)
+			return ret;
+	}
+
+	/* GPIO LDAC Initialization */
+	ret = no_os_gpio_get_optional(&dev->gpio_ldac, init_param->gpio_ldac_init);
+	if (ret)
+		return ret;
+
+	if (dev->gpio_ldac) {
+		return no_os_gpio_direction_output(dev->gpio_ldac,
+						   NO_OS_GPIO_HIGH);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Initialize registers based on init parameters.
+ * @param dev - The device structure.
+ * @param init_param - Pointer to structure containing the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_reg_init(struct ad5754r_dev* dev,
+		     struct ad5754r_init_param *init_param)
+{
+	int ret;
+	uint8_t chn;
+
+	if (!dev || !init_param)
+		return -EINVAL;
+
+	for (chn = 0; chn < AD5754R_NUM_CHANNELS; chn++) {
+		/* DAC Channel Range Select */
+		ret = ad5754r_set_ch_range(dev, chn, init_param->dac_ch_range[chn]);
+		if (ret)
+			return ret;
+
+		/* DAC Channel Powerup */
+		ret = ad5754r_set_ch_pwrup(dev, chn,
+					   init_param->dac_ch_pwr_states[chn]);
+		if (ret)
+			return ret;
+	}
+
+	/* Current Clamp Enable */
+	ret = ad5754r_set_current_clamp_en(dev, init_param->clamp_en);
+	if (ret)
+		return ret;
+
+	/* Clear Mode Setting */
+	ret = ad5754r_set_clear_mode(dev, init_param->clear_sel);
+	if (ret)
+		return ret;
+
+	/* SDO Disable */
+	ret = ad5754r_set_sdo_disable(dev, init_param->sdo_dis);
+	if (ret)
+		return ret;
+
+	/* Internal Reference Powerup */  // TODO figure out where to do this
+	return ad5754r_set_int_ref_pwrup(dev, init_param->int_ref_pwrup);
+}
+
+/**
+ * @brief Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - Pointer to structure containing the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_init(struct ad5754r_dev **device,
+		 struct ad5754r_init_param *init_param)
+{
+	struct ad5754r_dev *dev;
+	int ret;
+
+	if (!device || !init_param)
+		return -EINVAL;
+
+	dev = (struct ad5754r_dev *)no_os_malloc(sizeof(*dev));
+	if (!dev)
+		return -EINVAL;
+
+	/* GPIO Initialization */
+	ret = ad5754r_gpio_init(dev, init_param);
+	if (ret)
+		goto error_gpio;
+
+	/* SPI Initialization */
+	ret = no_os_spi_init(&dev->spi_desc, init_param->spi_init);
+	if (ret)
+		goto error_gpio;
+
+	/* Verify whether SPI is working by writing to a register
+	 * and reading it back */
+	ret = ad5754r_spi_verify(dev);
+	if (ret)
+		goto error_dev;
```

Comment:
you'd want here to remove spi, gpio and device.

Reply:
The error_dev (which points to `ad5754r_remove` )removes resources allocated to spi, gpio and the device

Reply:
the memory handling needs a bit of rework. See please previous comment.